File: electron/patcher/grf/reader.ts
import fs from 'fs';
import zlib from 'zlib';
import { GRF_HEADER_SIZE, GRF_SIGNATURE, GrfEntry, GrfHeader } from './types';

export class GrfReader {
    private filePath: string;
    private descriptor: number | null = null;

    constructor(filePath: string) {
        this.filePath = filePath;
    }

    public close() {
        if (this.descriptor !== null) {
            fs.closeSync(this.descriptor);
            this.descriptor = null;
        }
    }

    private open() {
        if (!fs.existsSync(this.filePath)) {
            throw new Error(`GRF file not found: ${this.filePath}`);
        }
        this.descriptor = fs.openSync(this.filePath, 'r');
    }

    public readHeader(): GrfHeader {
        if (this.descriptor === null) this.open();

        const buffer = Buffer.alloc(GRF_HEADER_SIZE);
        fs.readSync(this.descriptor!, buffer, 0, GRF_HEADER_SIZE, 0);

        // Signature Check (15 bytes)
        const signature = buffer.toString('latin1', 0, 15);
        if (signature !== GRF_SIGNATURE) {
            throw new Error('Invalid GRF signature');
        }

        // Key (14 bytes) - 15 to 29
        const key = buffer.subarray(15, 29);

        // File Table Offset (uint64) - 30 to 37
        // Note: JS numbers lose precision > 2^53, but GRF sizes usually fit.
        // We use BigInt just in case then convert.
        const fileTableOffset = Number(buffer.readBigUInt64LE(30));

        // Real File Count (uint32) - 38 to 41 (Actually seeds + counts in old versions, but for 2.0 structure...)
        // Wait, standard 2.0 structure:
        // 0-14: Sig
        // 15-29: Key
        // 30-37: FileTableOffset
        // 38-41: Seed (uint32) - often realFileCount - seed - 7 logic?
        // 42-45: RealFileCount (uint32)
        // Let's re-verify specific offset. 
        // Docs say: 46 bytes.
        // Offset 38 (4 bytes) = Seed
        // Offset 42 (4 bytes) = FilesCount (Raw)
        // RealFileCount = Raw - Seed - 7

        const seed = buffer.readInt32LE(38);
        const rawFileCount = buffer.readInt32LE(42);
        const realFileCount = rawFileCount - seed - 7;

        const version = 0x200; // Assuming 2.0 for this implementation

        return {
            signature,
            key,
            fileTableOffset,
            realFileCount,
            version
        };
    }

    public async readFileTable(header: GrfHeader): Promise<Map<string, GrfEntry>> {
        if (this.descriptor === null) this.open();

        // 1. Read Compressed Table Info (8 bytes: CompressedSize, RealSize)
        const tableInfoBuffer = Buffer.alloc(8);
        const absoluteTableOffset = header.fileTableOffset + GRF_HEADER_SIZE;

        fs.readSync(this.descriptor!, tableInfoBuffer, 0, 8, absoluteTableOffset);

        const compressedSize = tableInfoBuffer.readInt32LE(0);
        const realSize = tableInfoBuffer.readInt32LE(4);

        // 2. Read Compressed Data
        const compressedData = Buffer.alloc(compressedSize);
        fs.readSync(this.descriptor!, compressedData, 0, compressedSize, absoluteTableOffset + 8);

        // 3. Decompress
        const data = zlib.unzipSync(compressedData); // or inflateSync

        if (data.length !== realSize) {
            console.warn(`Warning: Decompressed table size mismatch. Expected ${realSize}, got ${data.length}`);
        }

        // 4. Parse Entries
        // Entry format: 
        // FileName (string, null terminated?)
        // CompressedSize (4)
        // CompressedSizeAligned (4)
        // RealSize (4)
        // Flags (1)
        // Offset (4)

        const entries = new Map<string, GrfEntry>();
        let offset = 0;
        let fileCount = 0;

        while (offset < data.length && fileCount < header.realFileCount) {
            // Read filename (variable length, null terminated generally or specific parsing logic)
            // Usually read until 0x00
            let endName = offset;
            while (endName < data.length && data[endName] !== 0) {
                endName++;
            }

            const filename = data.toString('latin1', offset, endName); // Encoding usually EUC-KR or ANSI, usage Latin1 preserves bytes
            offset = endName + 1; // Skip null

            const entryCompressedSize = data.readInt32LE(offset);
            offset += 4;
            const entryCompressedSizeAligned = data.readInt32LE(offset);
            offset += 4;
            const entryRealSize = data.readInt32LE(offset);
            offset += 4;
            const entryFlags = data.readUInt8(offset);
            offset += 1;
            const entryOffset = data.readInt32LE(offset);
            offset += 4;

            entries.set(filename.toLowerCase(), {
                filename,
                compressedSize: entryCompressedSize,
                compressedSizeAligned: entryCompressedSizeAligned,
                realSize: entryRealSize,
                flags: entryFlags,
                offset: entryOffset
            });

            fileCount++;
        }

        return entries;
    }
}

File: src-tauri/src/patcher/grf/reader.rs
use super::types::*;
use std::collections::HashMap;
use std::fs::File;
use std::io::{Read, Seek, SeekFrom};

pub struct GrfReader {
    file_path: String,
}

impl GrfReader {
    pub fn new(file_path: String) -> Self {
        Self { file_path }
    }

    pub fn read_header(&self) -> Result<GrfHeader, String> {
        let mut file = File::open(&self.file_path)
            .map_err(|e| format!("Failed to open GRF: {}", e))?;

        let mut buffer = vec![0u8; GRF_HEADER_SIZE];
        file.read_exact(&mut buffer)
            .map_err(|e| format!("Failed to read header: {}", e))?;

        // Signature (15 bytes)
        let signature = String::from_utf8_lossy(&buffer[0..15]).to_string();
        if signature != GRF_SIGNATURE {
            return Err("Invalid GRF signature".to_string());
        }

        // Key (14 bytes)
        let key = buffer[15..29].to_vec();

        // File Table Offset (8 bytes, little-endian u64)
        let file_table_offset = u64::from_le_bytes([
            buffer[30], buffer[31], buffer[32], buffer[33],
            buffer[34], buffer[35], buffer[36], buffer[37],
        ]);

        // Seed and file count
        let seed = i32::from_le_bytes([buffer[38], buffer[39], buffer[40], buffer[41]]);
        let raw_file_count = i32::from_le_bytes([buffer[42], buffer[43], buffer[44], buffer[45]]);
        let real_file_count = raw_file_count - seed - 7;

        Ok(GrfHeader {
            signature,
            key,
            file_table_offset,
            real_file_count,
            version: 0x200,
        })
    }

    pub fn read_file_table(&self, header: &GrfHeader) -> Result<HashMap<String, GrfEntry>, String> {
        let mut file = File::open(&self.file_path)
            .map_err(|e| format!("Failed to open GRF: {}", e))?;

        let absolute_table_offset = header.file_table_offset + GRF_HEADER_SIZE as u64;
        file.seek(SeekFrom::Start(absolute_table_offset))
            .map_err(|e| format!("Failed to seek to table: {}", e))?;

        // Read compressed table size and real size
        let mut table_info = vec![0u8; 8];
        file.read_exact(&mut table_info)
            .map_err(|e| format!("Failed to read table info: {}", e))?;

        let compressed_size = i32::from_le_bytes([table_info[0], table_info[1], table_info[2], table_info[3]]) as usize;
        let _real_size = i32::from_le_bytes([table_info[4], table_info[5], table_info[6], table_info[7]]);

        // Read compressed data
        let mut compressed_data = vec![0u8; compressed_size];
        file.read_exact(&mut compressed_data)
            .map_err(|e| format!("Failed to read compressed table: {}", e))?;

        // Decompress using flate2
        use flate2::read::ZlibDecoder;
        let mut decoder = ZlibDecoder::new(&compressed_data[..]);
        let mut data = Vec::new();
        decoder.read_to_end(&mut data)
            .map_err(|e| format!("Failed to decompress table: {}", e))?;

        // Parse entries
        let mut entries = HashMap::new();
        let mut offset = 0;
        let mut file_count = 0;

        while offset < data.len() && file_count < header.real_file_count {
            // Read filename (null-terminated)
            let mut end_name = offset;
            while end_name < data.len() && data[end_name] != 0 {
                end_name += 1;
            }

            let filename = String::from_utf8_lossy(&data[offset..end_name]).to_string();
            offset = end_name + 1;

            if offset + 17 > data.len() {
                break;
            }

            let entry_compressed_size = i32::from_le_bytes([data[offset], data[offset + 1], data[offset + 2], data[offset + 3]]);
            offset += 4;
            let entry_compressed_size_aligned = i32::from_le_bytes([data[offset], data[offset + 1], data[offset + 2], data[offset + 3]]);
            offset += 4;
            let entry_real_size = i32::from_le_bytes([data[offset], data[offset + 1], data[offset + 2], data[offset + 3]]);
            offset += 4;
            let entry_flags = data[offset];
            offset += 1;
            let entry_offset = i32::from_le_bytes([data[offset], data[offset + 1], data[offset + 2], data[offset + 3]]);
            offset += 4;

            entries.insert(
                filename.to_lowercase(),
                GrfEntry {
                    filename,
                    compressed_size: entry_compressed_size,
                    compressed_size_aligned: entry_compressed_size_aligned,
                    real_size: entry_real_size,
                    flags: entry_flags,
                    offset: entry_offset,
                    is_new: false,
                    data: None,
                },
            );

            file_count += 1;
        }

        Ok(entries)
    }
}
